### PR TITLE
Refactored UI code  and DISASM view massive bug fixed

### PR DESCRIPTION
**Pull request by maximilien-noal**
### Description of Changes

* fix: disasm view showed the wrong code, unless the window was closed and reopened
* feature: the disasm view now has the start address and number of instructions shown configurable
* refactor: the DebugViewModel and MainWindowViewModel are more SOLID
* refactor: the relationship between ViewModels are simpler
* refactor: and more UI code deleted / simplified
* split up and simplified xaml code
* Use of DataTemplates to enable lazy loading, so the Debug window loads faster (hopefully)
* Those DataTemplates will enable future features, such as multiple Memory and Disasm views opened at the same time

### Rationale behind Changes

Enable more contributions to the UI.
Makes the UI code easier to maintain.
Makes the implementation of future debug features easier.

### Suggested Testing Steps

Already tested, and I can't wait for people to test this. :)

### CODE REVIEW

File: src/Spice86/Infrastructure/IHostStorageProvider.cs [hunk 1]
```diff
@@ -57,4 +61,25 @@ public interface IHostStorageProvider {
...
+    /// </summary>
+    /// <param name="bitmap">The bitmap to save on the host storage.</param>
+    /// <returns>The operation as an awaitable Task.</returns>
+    Task SaveBitmapFile(WriteableBitmap bitmap);
+
+    /// <summary>
+    /// Spanws the file pciker to save the emulator state to a file.
+    /// </summary>
+    /// <param name="configuration">The emulator configuration, used for RecordedDataDirectory path property.</param>
+    /// <param name="programExecutor">The emulated program's executor, to save the</param>
```
Comment (JorisVanEijden) [inline src/Spice86/Infrastructure/IHostStorageProvider.cs:76]: to save the what? ;)

File: src/Spice86/ViewModels/DisassemblyViewModel.cs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        Decoder decoder = InitializeDecoder(codeReader, currentIp);
+        try {
+            DecodeFiftyInstructions(cpu, memory, emulatedMemoryStream, currentIp, decoder);
+        } finally {
+            emulatedMemoryStream.Dispose();
+        }
+    }
+
+    // TODO: Infinite scroll of instructions (UI paging)
+    private void DecodeFiftyInstructions(Cpu cpu, IMemory memory, EmulatedMemoryStream emulatedMemoryStream, uint currentIp,
```
Comment (JorisVanEijden) [inline src/Spice86/ViewModels/DisassemblyViewModel.cs:110]: You probably want to make the amount of decoded instructions configurable

File: src/Spice86/ViewModels/MainWindowViewModel.cs [hunk 3]
```diff
@@ -30,6 +30,7 @@
 using Avalonia.Media.Imaging;
 using Avalonia.Platform;
 
+using Spice86.Core.Emulator.InternalDebugger;
```
Comment (JorisVanEijden) [inline src/Spice86/ViewModels/MainWindowViewModel.cs:33]: This does not appear to be used?

File: src/Spice86/ViewModels/MainWindowViewModel.cs [hunk 4]
```diff
@@ -42,14 +43,15 @@
 public sealed partial class MainWindowViewModel : ViewModelBase, IPauseStatus, IGui, IDisposable {
     private const double ScreenRefreshHz = 60;
     private readonly ILoggerService _loggerService;
-    private readonly IUIDispatcherTimer _uiDispatcherTimer;
+    private readonly IUIDispatcherTimerFactory _iuiDispatcherTimerFactory;
```
Comment (JorisVanEijden) [inline src/Spice86/ViewModels/MainWindowViewModel.cs:46]:
```suggestion
    private readonly IUIDispatcherTimerFactory _uiDispatcherTimerFactory;
```

File: src/Spice86/ViewModels/MainWindowViewModel.cs [hunk 5]
```diff
@@ -73,17 +75,17 @@ public sealed partial class MainWindowViewModel : ViewModelBase, IPauseStatus, I
 
     public ITimeMultiplier? ProgrammableIntervalTimer { private get; set; }
 
-    public MainWindowViewModel(IAvaloniaKeyScanCodeConverter avaloniaKeyScanCodeConverter, IProgramExecutorFactory programExecutorFactory, IWindowActivator windowActivator, IUIDispatcher uiDispatcher, IHostStorageProvider hostStorageProvider, ITextClipboard textClipboard, IUIDispatcherTimer uiDispatcherTimer, Configuration configuration, ILoggerService loggerService) {
+    public MainWindowViewModel(IAvaloniaKeyScanCodeConverter avaloniaKeyScanCodeConverter, IProgramExecutorFactory programExecutorFactory, IDebugWindowActivator debugWindowActivator, IUIDispatcher uiDispatcher, IHostStorageProvider hostStorageProvider, ITextClipboard textClipboard, IUIDispatcherTimerFactory iuiDispatcherTimerFactory, Configuration configuration, ILoggerService loggerService) {
```
Comment (JorisVanEijden) [inline src/Spice86/ViewModels/MainWindowViewModel.cs:78]:
```suggestion
    public MainWindowViewModel(IAvaloniaKeyScanCodeConverter avaloniaKeyScanCodeConverter, IProgramExecutorFactory programExecutorFactory, IDebugWindowActivator debugWindowActivator, IUIDispatcher uiDispatcher, IHostStorageProvider hostStorageProvider, ITextClipboard textClipboard, IUIDispatcherTimerFactory uiDispatcherTimerFactory, Configuration configuration, ILoggerService loggerService) {
```

File: src/Spice86/ViewModels/SoftwareMixerViewModel.cs [hunk 6]
```diff
@@ -1,42 +1,32 @@
 namespace Spice86.ViewModels;
 
 using Avalonia.Collections;
-using Avalonia.Threading;
+using Avalonia.Controls;
```
Comment (JorisVanEijden) [inline src/Spice86/ViewModels/SoftwareMixerViewModel.cs:4]: This does not appear to be used?

File: src/Spice86/ViewModels/VideoCardViewModel.cs [hunk 7]
```diff
@@ -1,255 +1,35 @@
...
-using Spice86.Core.Emulator.CPU;
-using Spice86.Core.Emulator.Devices.Sound;
-using Spice86.Core.Emulator.Devices.Sound.Midi;
 using Spice86.Core.Emulator.Devices.Video;
 using Spice86.Core.Emulator.InternalDebugger;
-using Spice86.Core.Emulator.Memory;
-using Spice86.Infrastructure;
-using Spice86.Interfaces;
-using Spice86.MemoryWrappers;
+using Spice86.Core.Emulator.VM;
```
Comment (JorisVanEijden) [inline src/Spice86/ViewModels/VideoCardViewModel.cs:7]: This does not appear to be used?

**Review by JorisVanEijden — CHANGES_REQUESTED** [review]
Seems to work fine. No major issue. Just a few nitpicks.

File: src/Spice86/Infrastructure/IHostStorageProvider.cs (hunk 1 above)
Comment (maximilien-noal) [reply src/Spice86/Infrastructure/IHostStorageProvider.cs]: Corrected.

File: src/Spice86/ViewModels/MainWindowViewModel.cs (hunk 3 above)
Comment (maximilien-noal) [reply src/Spice86/ViewModels/MainWindowViewModel.cs:33]: Corrected.

File: src/Spice86/ViewModels/DisassemblyViewModel.cs (hunk 2 above)
Comment (maximilien-noal) [reply src/Spice86/ViewModels/DisassemblyViewModel.cs]: Done. The start address is now configurable, too.

File: src/Spice86/ViewModels/DebugWindowViewModel.cs [hunk 8]
```diff
@@ -61,11 +66,34 @@ public partial class DebugWindowViewModel : ViewModelBase, IInternalDebugger {
...
+    
+    [RelayCommand]
+    public void Pause() {
+        if (_programExecutor is null || _pauseStatus is null) {
+            return;
+        }
+        _pauseStatus.IsPaused = _programExecutor.IsPaused = IsPaused = true;
+    }
+
+    [RelayCommand(CanExecute = nameof(IsPaused))]
```
Comment (JorisVanEijden) [inline src/Spice86/ViewModels/DebugWindowViewModel.cs:80]: `IsPaused ` does not exist in this class, so the Continue button does not work anymore.

**Review by JorisVanEijden — CHANGES_REQUESTED** [review]
Last commit broke the Continue button

File: src/Spice86/ViewModels/DebugWindowViewModel.cs (hunk 8 above)
Comment (maximilien-noal) [reply src/Spice86/ViewModels/DebugWindowViewModel.cs:80]: The IsPaused property is source generated by the ObservableProperty attribute.

What was missing on the _isPaused field, was:

    [NotifyCanExecuteChangedFor(nameof(ContinueCommand))]

**Comment by maximilien-noal** [comment]
I fixed the last commit. :)

Thank you for your review!

**Review by JorisVanEijden — APPROVED** [review]

**Pull request merged**